### PR TITLE
Dijkstra

### PR DESCRIPTION
**Pull request by JHLee0513**
Write a detailed description about the code you want to merge.

--------------------------------------------------------------
Checklist:
- [x] Tests
- [x] Documentation/Header comments
- [x] Linting

### CODE REVIEW

File: include/planner/Dijkstra.hpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#include "planner/Planner.hpp"
+#include "distance/distance.hpp"
+
+namespace libcozmo {
+namespace planner {
+
+/// This class uses Dijkstra's algorithm to find a sequence of actions
+/// to get from start state to goal state
+///
+/// To search for a path the class implements an actionspace, a statespace,
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: Remove this comment

File: include/planner/Dijkstra.hpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+/// and a model
+class Dijkstra : public virtual Planner {
+ public:
+    /// Constructs a planner with given actionspace, statespace, model
+    ///
+    /// Constructs distance metric and goal tolerance that defines the solver's
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: Constructs distance metric and goal tolerance

File: include/planner/Dijkstra.hpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    /// additional stopping condition.
+    ///
+    /// \param actionspace The actionspace to get set of actions to explore
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: Defines the set of all possible actions

File: include/planner/Dijkstra.hpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    /// \param statespace The statespace to build graph representation with
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: Defines the set of all possible states

File: include/planner/Dijkstra.hpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    /// \param model Model to generate successors
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: A model that has learnt f: (state, action) -> state'

File: include/planner/Dijkstra.hpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    /// \param distance_metric The metric to calculate distance to goal
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: The metric used to calculate the distance between two states

File: include/planner/Dijkstra.hpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    /// \param goal_tolerance Additional stopping condition for solver
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: The threshold for being considered at the goal

File: include/planner/Dijkstra.hpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    Dijkstra(
+        libcozmo::actionspace::ActionSpace* actionspace,
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: shared_ptr

File: include/planner/Dijkstra.hpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        libcozmo::statespace::StateSpace* statespace,
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: shared_ptr

File: include/planner/Dijkstra.hpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        libcozmo::model::Model* model,
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: shared_ptr

File: include/planner/Dijkstra.hpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        distance::Distance* distance_metric,
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: shared_ptr

File: include/planner/Dijkstra.hpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        const double& goal_tolerance) : \
+        m_action_space(actionspace),
+        m_state_space(statespace),
+        m_model(model),
+        m_continuous_statespace(std::make_shared<aikido::statespace::SE2>()),
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: You should be passing in a shared_ptr, not making one in the constructor

File: include/planner/Dijkstra.hpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    bool set_goal(const int& goal_id) override;
+
+    /// Documentation inherited
+    bool solve(std::vector<int>* actions) override;
+
+ private:
+    /// Extracts sequence of actions from start to goal
+    ///
+    /// \param[out] actions Vector of action IDs
+    void extract_sequence(std::vector<int> *actions);
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: extract_action_sequence

File: include/planner/Dijkstra.hpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+    /// Finds all successor states given input state
+    ///
+    /// \param state_ The current state (in continuous space)
+    /// \param[out] successors Vector of succesor states
+    void get_successors(
+        const statespace::SE2::State* state_,
+        std::vector<statespace::SE2::State>* succesors);
+
+    /// Check whether the solver has reached the goal, which its defintion
```
Comment (vinitha910) [inline include/planner/Dijkstra.hpp]: Checks whether the solver has reached the goal (i.e. is within the specified goal tolerance)

File: include/planner/Planner.hpp [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+#include <utility>
+#include <unordered_map>
+
+namespace libcozmo {
+namespace planner {
+
+/// Maps state ID to it's appropiate cost
+typedef std::unordered_map<int, double> CostMap;
+
+/// Maps the child/successor state ID to its parent ID and to the action id
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: Maps the child/successor state ID to its respective (parent ID, action ID) pair

File: include/planner/Planner.hpp [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+/// that is taken to get to the child state from parent state
+typedef std::unordered_map<int, std::pair<
+    int, int>> ChildToParentMap;
+
+// Comparator orders state IDs based on their costs from start
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: ...costs-to-come from the start state

File: include/planner/Planner.hpp [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    bool operator()(const int& state_1, const int& state_2) const {
+        return cost_map_.find(state_1)->second <=
+        cost_map_.find(state_2)->second;
+    }
+
+ private:
+    const CostMap& cost_map_;
+};
+
+/// This class plans a sequence of actions required reach from given start to
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: required to get from the start to goal state

File: include/planner/Planner.hpp [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+/// goal state. The strategy to which the solution is found will vary based on
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: The planning method varies based on the derived class.

File: include/planner/Planner.hpp [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+/// the derived classes.
+///
+/// For finding a solution the class utilizes an actionspace to get set of
+/// actions to explore, statespace to build a graph representation, and
+/// a model to get successors.
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: to generate successors

File: include/planner/Planner.hpp [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+ private:
+    const CostMap& cost_map_;
+};
+
+/// This class plans a sequence of actions required reach from given start to
+/// goal state. The strategy to which the solution is found will vary based on
+/// the derived classes.
+///
+/// For finding a solution the class utilizes an actionspace to get set of
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: While searching, the class...

File: include/planner/Planner.hpp [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    /// \return true if start_id valid; false otherwise
+    virtual bool set_start(const int& start_id) = 0;
+
+    /// Sets goal ID
+    ///
+    /// \param goal_id The goal state ID
+    /// \return true if goal_id valid; false otherwise
+    virtual bool set_goal(const int& goal_id) = 0;
+
+    /// Solves given search problem and finds a sequence of actions from the
```
Comment (vinitha910) [inline include/planner/Planner.hpp]: Find a sequence of actions from the start to goal state

File: src/planner/Dijkstra.cpp [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+
+#include "planner/Dijkstra.hpp"
+#include <iostream>
+
+namespace libcozmo {
+namespace planner {
+
+    bool Dijkstra::set_start(const int& start_id) {
+        if (m_state_space->is_valid_state(
+                *m_state_space->get_state(start_id))) {
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp]: This will segfault if `m_state_space->get_state(start_id)` returns a `nullptr`

File: src/planner/Dijkstra.cpp [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+            m_start_id = start_id;
+            return true;
+        }
+        return false;
+    }
+
+    bool Dijkstra::set_goal(const int& goal_id) {
+        if (m_state_space->is_valid_state(
+                *m_state_space->get_state(goal_id))) {
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp]: This will segfault if `m_state_space->get_state(goal_id)` returns a `nullptr`

File: src/planner/Dijkstra.cpp [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+            m_goal_id = goal_id;
+            return true;
+        }
+        return false;
+    }
+
+    bool Dijkstra::solve(std::vector<int>* actions) {
+        if (m_goal_id == -1 || m_start_id == -1) {
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:63]: // Check that goal and start ID have set.

File: src/planner/Dijkstra.cpp [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+            return false;
+        }
+        CostMap costmap;
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:68]: newline
// Initializing the queue and cost map

File: src/planner/Dijkstra.cpp [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+        CostMapComparator comparator(costmap);
+        std::set<int, CostMapComparator> Q(comparator);
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp]: `priority_queue`

File: src/planner/Dijkstra.cpp [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+        std::vector<int> path_ids;
+        Q.insert(m_start_id);
+        costmap[m_start_id] = 0;
+        while (!Q.empty()) {
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp]: newline
// While the priority queue in not empty, search for a path from the start state to the goal state

Comment (vinitha910) [reply src/planner/Dijkstra.cpp]: Add more comments like this and line breaks through out this while loop

File: src/planner/Dijkstra.cpp [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+
+    bool Dijkstra::is_goal(const int& curr_state_id) const {
+        return curr_state_id  == m_goal_id || m_distance_metric->get_distance(
+            *m_state_space->get_state(curr_state_id),
+            *m_state_space->get_state(m_goal_id)) <= m_goal_tolerance;
+    }
+
+    void Dijkstra::get_successors(
+        const statespace::SE2::State* curr_state,
+        std::vector<statespace::SE2::State>* succesors) {
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:136]: Add spaces between lines and comments where appropiate

**Review by vinitha910 — CHANGES_REQUESTED** [review]

File: include/planner/Planner.hpp (hunk 19 above)
Comment (JHLee0513) [reply include/planner/Planner.hpp]: additional comment: talk about how planner handles all the states using its ID, and the conversion between continuous to discrete etc here.

File: src/planner/Dijkstra.cpp [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+                if (m_state_space->is_valid_state(succesor_state)) {
+                    const int succesor_id =
+                        m_state_space->get_or_create_state(succesor_state);
+                    const double new_cost = costmap[curr_state_id] +
+                        m_state_space->get_distance(
+                            *curr_state_,
+                            succesor_state);
+
+                    // Update the cost map if a successor state has not been
+                    // explored before, or if it has lower cost than previously
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:106]: has a lower cost than that in the priority queue

File: src/planner/Dijkstra.cpp [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+                    // explored path
+                    if (costmap.find(succesor_id) == costmap.end() ||
+                        costmap[succesor_id] > new_cost) {
+
+                        // Add (parent ID, action ID) pair to child to parent
+                        // map if the explored succsor is being added
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:112]: Map successor ID -> (parent ID, action ID)

File: src/planner/Dijkstra.cpp [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+        const statespace::SE2::State* curr_state,
+        std::vector<statespace::SE2::State>* succesors) {
+        for (int i = 0; i < m_action_space->size(); i++) {
+            const actionspace::ActionSpace::Action* action =
+                m_action_space->get_action(i);
+            const auto action_ = static_cast<
+                const actionspace::GenericActionSpace::Action*>(action);
+            
+            // Convert current state and action into model input
+            const model::DeterministicModel::DeterministicModelInput input(
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:144]: Should take generic model (you can make this change once I merge the model PR)

File: src/planner/Dijkstra.cpp [hunk 32]
```diff
@@ -0,0 +1,186 @@
...
+                *action_);
+            model::DeterministicModel::DeterministicModelOutput output;
+            
+            // Given model input, get successor state in continuous space
+            m_model->get_prediction(input, &output);
```
Comment (vinitha910) [inline src/planner/Dijkstra.cpp:149]: hmm, I think I'm going to have the model output the new state

**Review by vinitha910 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**